trieste/acquisition/function.py

```diff
@@ -326,47 +410,85 @@ def prepare_acquisition_function(
 
-    :param model: The model of the objective function.
-    :param samples: Samples from the distribution over :math:`y^*`.
-    :return: The max-value entropy search acquisition function modified for objective
-        minimisation. This function will raise :exc:`ValueError` or
-        :exc:`~tf.errors.InvalidArgumentError` if used with a batch size greater than one.
-    :raise ValueError or tf.errors.InvalidArgumentError: If ``samples`` has rank less than two, or
-        is empty.
-    """
-    tf.debugging.assert_rank(samples, 2)
-    tf.debugging.assert_positive(len(samples))
 
-    def acquisition(x: TensorType) -> TensorType:
+class min_value_entropy_search:
+    def __init__(self, model: ProbabilisticModel, samples: TensorType):
+        r"""
+        Return the max-value entropy search acquisition function (adapted from :cite:`wang2017max`),
+        modified for objective minimisation. This function calculates the information gain (or
+        change in entropy) in the distribution over the objective minimum :math:`y^*`, if we were
+        to evaluate the objective at a given point.
+
+        :param model: The model of the objective function.
+        :param samples: Samples from the distribution over :math:`y^*`.
+        :return: The max-value entropy search acquisition function modified for objective
+            minimisation. This function will raise :exc:`ValueError` or
+            :exc:`~tf.errors.InvalidArgumentError` if used with a batch size greater than one.
+        :raise ValueError or tf.errors.InvalidArgumentError: If ``samples`` has rank less than two,
+            or is empty.
+        """
+        tf.debugging.assert_rank(samples, 2)
+        tf.debugging.assert_positive(len(samples))
+
+        self._model = model
+        self._samples = tf.Variable(samples)
+
+    def update(self, samples: TensorType) -> None:
+        """Update the acquisition function with new samples."""
+        self._samples.assign(samples)
+
+    @tf.function
+    def __call__(self, x: TensorType) -> TensorType:
         tf.debugging.assert_shapes(
             [(x, [..., 1, None])],
             message="This acquisition function only supports batch sizes of one.",
         )
-        fmean, fvar = model.predict(tf.squeeze(x, -2))
+        fmean, fvar = self._model.predict(tf.squeeze(x, -2))
         fsd = tf.math.sqrt(fvar)
         fsd = tf.clip_by_value(
             fsd, CLAMP_LB, fmean.dtype.max
         )  # clip below to improve numerical stability
 
         normal = tfp.distributions.Normal(tf.cast(0, fmean.dtype), tf.cast(1, fmean.dtype))
-        gamma = (tf.squeeze(samples) - fmean) / fsd
+        gamma = (tf.squeeze(self._samples) - fmean) / fsd
 
         log_minus_cdf = normal.log_cdf(-gamma)
         ratio = tf.math.exp(normal.log_prob(gamma) - log_minus_cdf)
         f_acqu_x = -gamma * ratio / 2 - log_minus_cdf
 
         return tf.math.reduce_mean(f_acqu_x, axis=1, keepdims=True)
 
-    return acquisition
-
```

Comment:
How come we dont have a new lower_confidence_bound class (its still a function like before). Just because it doesnt have anything to update doesnt mean we dont want to stop recompiling it?

Reply:
Because I haven't updated all the acquisition functions yet. See the text box for the PR.